File: speedtests/go/vec.go
// Set of wrappers for gonum vector algebra

package speedtests

import (
	"fmt"
	"gonum.org/v1/gonum/mat"
	"math/rand"
	"time"
)

const (
	VN = 1 << (iota * 10) // Sizes of vectors
	VNs
	VNm
	VNl
)

// createRandomVector create random N vector
func createRandomVector(N int) *mat.VecDense {
	v := mat.NewVecDense(N, nil)
	for i := 0; i < N; i++ {
		rand.Seed(int64(time.Now().UnixNano()))
		v.SetVec(i, rand.Float64() * 100)
	}
	return v
}

// scaleVector scale vector v
func scaleVector(v *mat.VecDense, alpha float64) {
	v.ScaleVec(alpha, v)
}

// frobeniusNormOfVector return Frobenius norm of vector
func frobeniusNormOfVector(v *mat.VecDense) (f float64) {
	f = mat.Norm(v, 2)
	return
}

// additionOfVectors return sum of vectors v, u,
// if alpha is not 0 return scaled sum of vectors (v + alpha * u)
func additionOfVectors(v, u *mat.VecDense, alpha float64) (w *mat.VecDense) {
	w = mat.NewVecDense(v.Len(), nil)
	if alpha == 0.0 {
		w.AddVec(u, v)
	}
	if alpha != 0.0 {
		w.AddScaledVec(v, alpha, u)
	}
	return
}

// subtractOfVectors return subtract of vectors v, u
func subtractOfVectors(v, u *mat.VecDense) (w *mat.VecDense) {
	w = mat.NewVecDense(v.Len(), nil)
	w.SubVec(v, u)
	return
}

// dotOfVectors return result of multiplication of two vectors
func dotOfVectors(v, u *mat.VecDense) (w float64) {
	w = mat.Dot(v, u)
	return
}

// vecPrint print vector in stdout
func vecPrint(v *mat.VecDense) {
	fmt.Printf("%v\n",
		mat.Formatted(v, mat.Prefix(" "), mat.Excerpt(3)))
}

File: speedtests/python/vec.py
# Testing speed of operations
# For every size from: (1024, 16384, 65536, 131072, 262144, 524288):
# - create random vectors [0;100);
# - scale vector;
# - Frobenius norm of vector;
# - addition of vectors;
# - subtract of vectors;
# - dot of vectors.

import os
import time
import traceback
import numpy as np
from numpy import linalg as LA
from memory_profiler import profile

DEFAULT = 10
try:
    count = os.environ["PYTHON_TESTS_COUNT"]
except KeyError:
    count = DEFAULT
except Exception as e:
    print(traceback.format_exc())
    raise e


def timer(count):
    """
    Таймер для замера скорости работы тестируемых функций.
    :param count: int - кол-во прогонов функции.
    """

    def wrp(func):
        def wrapper(*args, **kwargs):
            start = time.time()
            for n in range(count):
                value = func(*args, **kwargs)
            end = time.time()
            runtime = (end - start)/count
            print(f"\nTime of work {func.__name__}({args}): {runtime}\n")
            return value
        # декоратор profile - изменяет расход памяти исполняемым кодом.
        return profile(precision=10)(wrapper)
    return wrp


@timer(count=count)
def create_random_vector(n: int) -> np.ndarray:
    """
    Создает вектор с рандомными элементами [0;100)
    :param n: int - shape.
    :return:
    - <class 'numpy.ndarray'> - рандомный вектор.
    """

    return np.random.rand(n) * 100


@timer(count=count)
def scale_vector(vec: np.ndarray, alpha: float):
    """
    Масштабирует переданный вектор на alpha.
    :param vec: np.ndarray - вектор для масштабирования;
    :param alpha: float - параметр масштабирования.
    """

    vec * alpha


@timer(count=count)
def frobenius_norm_of_vector(vec: np.ndarray) -> float:
    """
    Считает норму Фробениуса для переданного вектора.
    :param vec: np.ndarray - вектор для поиска нормы;
    :return:
    - float - норма Фробениуса переданного вектора.
    """

    return LA.norm(vec)


@timer(count=count)
def addition_of_vectors(vec: np.ndarray, vec_2: np.ndarray) -> np.ndarray:
    """
    Считает сумму двух векторов.
    :param vec: np.ndarray - слагаемое;
    :param vec_2: np.ndarray - слагаемое.
    :return:
    - np.ndarray - сумма.
    """

    return np.add(vec, vec_2)


@timer(count=count)
def subtract_of_vectors(vec: np.ndarray, vec_2: np.ndarray) -> np.ndarray:
    """
    Считает разницу двух векторов.
    :param vec: np.ndarray - уменьшаемое;
    :param vec_2: np.ndarray - вычитаемое.
    :return:
    - np.ndarray - разность.
    """

    return np.subtract(vec, vec_2)


@timer(count=count)
def dot_of_vectors(vec: np.ndarray, vec_2: np.ndarray) -> np.ndarray:
    """
    Считает произведение двух векторов.
    :param vec: np.ndarray - множитель;
    :param vec_2: np.ndarray - множитель.
    :return:
    - np.ndarray - произведение.
    """

    return np.dot(vec, vec_2)
